gossipsub.go

```diff
@@ -209,20 +250,28 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 }
 
 func (gs *GossipSubRouter) Publish(from peer.ID, msg *pb.Message) {
-	gs.mcache.Put(msg)
-
-	tosend := make(map[peer.ID]struct{})
 	for _, topic := range msg.GetTopicIDs() {
+		config, ok := gs.topicConfigs[topic]
+		if ok {
+			config.Publish(gs, from, msg)
+		}
+	}
+}
+
+func (gs *GossipSubRouter) AddGossipPeers(tosend map[peer.ID]struct{}, topics []string, withFloodSubPeers bool) {
+	for _, topic := range topics {
 		// any peers in the topic?
 		tmap, ok := gs.p.topics[topic]
 		if !ok {
 			continue
```

Comment:
@vyzo is it intentional that during a `Publish` we do not initialize `gs.fanout[topic]` if there are no peers in the topic? By not doing this we ensure that we will not rebroadcast (via emitGossip) our published message even once we come across new peers. This is less relevant in the current implementation since the republishing window is relatively short (5 heartbeat intervals which is 500ms), however it might still be worth correcting this even for our existing implementation.

Reply:
They won't be added to fanout unless there is a new message published.
Also, the heartbeat is 1 _second_, so 5 intervals are 5s.

Reply:
My bad, was looking at heartbeat initial delay instead of interval. Ok, so the following sequence of events in the current implementation is intentional?

**Scenario 1:**
1. A publishes M to GossipSub topic T, but has no peers
2. B subscribes to T
3. B connects to A
4. Nothing happens, B never receives M (even if within the cache history period)

**Scenario 2:**
1. A publishes M to GossipSub topic T, but has no peers
2. B1...B20 subscribes to T (important that it's more than `GossipSubD=6`, as mentioned [above](https://github.com/libp2p/go-libp2p-pubsub/pull/171#discussion_r273141427))
3. B1...B20 connects to A
4. Some of the B's receive M and some do not

**Scenario 3:**
1. A publishes M to GossipSub topic T, but has no peers
2. Wait 6 seconds
2. B1...B20 subscribes to T
3. B1...B20 connects to A
4. Nothing happens, none of the B's receives M

Reply:
Yes, that's fine.

Reply:
That psudo-use of caching seems pretty suspect to me, but that's fine I can preserve that behavior and just modify the persistence version.

Reply:
It's not designed to be a reliability mechanism; it's designed to help the network connect when the mesh is not fully connected for one reason or the other.